core-bundle/src/EventListener/PreviewToolbarListener.php

```diff
@@ -0,0 +1,203 @@
+
+namespace Contao\CoreBundle\EventListener;
+
+use Contao\ArticleModel;
+use Contao\BackendUser;
+use Contao\CoreBundle\Event\PreviewUrlConvertEvent;
+use Contao\CoreBundle\Exception\RedirectResponseException;
+use Contao\CoreBundle\Routing\ScopeMatcher;
+use Contao\CoreBundle\Security\Authentication\FrontendPreviewAuthenticator;
+use Contao\CoreBundle\Security\Authentication\Token\TokenChecker;
+use Contao\PageModel;
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpKernel\Event\ResponseEvent;
+use Symfony\Component\Routing\RouterInterface;
+use Symfony\Component\Security\Core\Authentication\Token\Storage\TokenStorageInterface;
+use Symfony\Component\Security\Core\Authorization\AuthorizationChecker;
+use Symfony\Component\Security\Csrf\CsrfTokenManagerInterface;
+use Twig\Environment as TwigEnvironment;
+use Twig\Error\Error as TwigError;
+
+/**
+ * PreviewToolbarListener injects the back end preview switch toolbar.
+ *
+ * The onKernelResponse method must be connected to the kernel.response event.
+ *
+ * The toolbar is only injected on well-formed HTML (with a proper <body> tag).
+ * This means that the WDT is never included in sub-requests or ESI requests.
+ */
+final class PreviewToolbarListener
+{
+
+    private $scopeMatcher;
+
+    private $previewScript;
+
+    private $twig;
+
+    private $tokenChecker;
+
+    private $router;
+
+    private $tokenManager;
+
+    private $csrfTokenName;
+
+    private $dispatcher;
+
+    private $frontendPreviewAuthenticator;
+
+    public function __construct(
+        string $previewScript,
+        ScopeMatcher $scopeMatcher,
+        TwigEnvironment $twig,
+        TokenChecker $tokenChecker,
+        RouterInterface $router,
+        CsrfTokenManagerInterface $tokenManager,
+        string $csrfTokenName,
+        EventDispatcherInterface $dispatcher, FrontendPreviewAuthenticator $frontendPreviewAuthenticator
+    ) {
+        $this->previewScript        = $previewScript;
+        $this->scopeMatcher         = $scopeMatcher;
+        $this->twig                 = $twig;
+        $this->tokenChecker         = $tokenChecker;
+        $this->router               = $router;
+        $this->tokenManager         = $tokenManager;
+        $this->csrfTokenName        = $csrfTokenName;
+        $this->dispatcher           = $dispatcher;
+        $this->frontendPreviewAuthenticator = $frontendPreviewAuthenticator;
+    }
+
+    public function onKernelResponse(ResponseEvent $event): void
+    {
+        if (!$this->scopeMatcher->isContaoMasterRequest($event)) {
+            return;
+        }
+
+        $request  = $event->getRequest();
+        $response = $event->getResponse();
+
+        if ($request->getScriptName() !== $this->previewScript) {
+            return;
+        }
+
+        // Ignore redirects and errors
+        if (200 !== $response->getStatusCode()) {
```

Comment:
Maybe this should use `$response->isSuccessful()` or at least `$response->isOk()`